docs/Custom Objects.md

```diff
@@ -0,0 +1,127 @@
+# Creating and initializing custom objects #
+
+In addition to using the engine's built-in objects, you will create,
+initialize, and use your own JS objects.
+This is especially true if you are using the JS engine with scripts to
+automate your application.
+Custom JS objects can provide direct program services, or they can serve
+as interfaces to your program's services.
+For example, a custom JS object that provides direct service might be
+one that handles all of an application's network access, or might serve
+as an intermediary broker of database services.
+Or a JS object that mirrors data and functions that already exist in the
+application may provide an object-oriented interface to C code that is
+not otherwise, strictly-speaking, object-oriented itself.
+Such a custom object acts as an interface to the application itself,
+passing values from the application to the user, and receiving and
+processing user input before returning it to the application.
+Such an object might also be used to provide access control to the
+underlying functions of the application.
+
+There are two ways to create custom objects that the JS engine can use:
+
+- Write a JS script that creates an object, its properties, methods, and
+  constructor, and then pass the script to the JS engine at run time.
+- Embed code in your application that defines the object's properties
+  and methods, call the engine to initialize a new object, and then set
+  the object's properties through additional engine calls.
+  An advantage of this method is that your application can contain
+  native methods that directly manipulate the object embedding.
+
+In either case, if you create an object and then want it to persist in
+the run time where it can be used by other scripts, you must root the
+object by calling `JS_AddRoot` or `JS_AddNamedRoot`.
```

Comment:
Oops, this does not apply anymore.

Reply:
Yeah, you could probably swap in JS::PersistentRooted<T> for these. Though it makes me nervous to recommend for much of anything; 90% of the uses of PersistentRooted end up creating unwanted leaks. It would almost be better to have a single PersistentRooted<MyGlobalStuff>, and have MyGlobalStuff be a struct that holds everything you need (with a trace() method that keeps it all alive.)

But anyway, for now it'd probably be fine to just talk about PersistentRooted here.

Reply:
...or even to suggest installing these sorts of things as named properties on the global, and not using PersistentRooted at all, perhaps?

Reply:
In GJS we use `JS_AddExtraGCRootsTracer` to trace things from the context, is that a reasonable approach?

Reply:
Yeah, that's totally reasonable and is probably more straightforward for many embeddings -- "do this to identify the embedding roots" as opposed to "this structure contains the embedding roots". I think both can make sense, it's dependent on the situation.